### ISSUE TITLE
Feature Request: Upload / Download Slots

### ISSUE THREAD

**Issue by jackBonadies** [jpdillingham/Soulseek.NET#633]
Hi JP,

I have a potential feature request for upload and download slots.  I don't think there is a way to do this currently.  I do have a proposed solution, so let me know if it sounds reasonable (if such a feature would be one you want in the library).

My proposed solution (for example upload slots) is to add two clauses, an entry clause right after awaiting the semaphore and an exit clause right before releasing the semaphore.  There would also be an additional member counter that gets incremented each time a new user queues files and decremented when they have downloaded all of their queued files, and a member dictionary similar to UploadSemaphores (maybe UploadGatekeeper) where the key is a user and the value is a tuple<AutoResetEvent, int, int> where the last two ints are tokens (i.e. GetNextToken()) the first is the time the user first queued files and the second is the last time they were able to download (these will be used in order to implement FIFO and Round Robin queues - i.e. ARE.Set() the user with minimum first int for FIFO or ARE.Set() the user with minimum second int for RR).  

In the entry clause, we do UploadGatekeeper.GetOrAdd(username, new Tuple<ARE, int,int>(new ARE(false), getToken(), -1)).  Then we check if we can proceed immediately (i.e. the current number of users is <= number slots), and if so continue, if we cannot proceed immediately we do ARE.WaitAsync().  

In the exit clause, we check if (current number of users is <= number slots).  If true, we don't do anything, no one is waiting on any events.  If false, we set a slot number of AREs.  We can take a user delegate to decide which ones to set, or set based on FIFO and RR options.

Thoughts?  Is there a better way to do this or one you had in mind?  If it would be valuable I can make implement said changes, or if you want to do the implementation feel free to consider the above or scrap the above.

Thanks!

**Comment by jpdillingham** [comment]
Thanks for your thoughts on this.  I put a lot of thought into this in the past (probably over a year ago now) and the public API for the library ended up being really messy, and still fairly constrictive for calling clients.  My conclusion was that the library wouldn't do any traffic control, outside of the single upload semaphore, and instead delegate to calling code.

I don't think what you've proposed is necessarily wrong, but relying on the internal async/await state machine(s) to represent the state of the queue prevents you from being able to recover from an application crash/restart in a predictable way.  This may not be a concern to everyone, since remote clients will re-request files when they detect that a users from which they are attempting to download has gone offline and come back.  The official clients don't preserve queue order through a restart, so it's not a big deal.

For my purposes (within slskd, not this library) my goal is to implement per-user and per-group quotas (daily, weekly, monthly), speed limits, caps on enqueued file count and size and a limit on how long files can remain enqueued.  All of this requires durable storage and ordering of enqueued downloads (as well as history, but that's tangential to your proposal).

My goal has been to keep as much state out of the library as possible, with exceptions being (at least, the ones that are top of mind) the message connection cache, state for the distributed network, and the semaphore that prevents concurrent uploads to a single user.

I'll keep thinking about a way to expose an interface that would simplify queue management so implementing code doesn't have to roll their own.  I think it would be good to solve for at least speed limits and number of total concurrent uploads.

**Comment by jpdillingham** [comment]
I've added a very basic queue in #640, but I'll leave this issue open because there's a lot that can be approved.

Awaiting a task after acquiring the per-user semaphore is an effective way to control uploads across users, so I think a lot of the logic you suggest would work.  I'm continuing to think of a clean way to implement this.

**Comment by jpdillingham** [comment]
I'm thinking of adding an additional line of code just before the code that awaits the upload slot semaphore that essentially amounts to:

```c#
await UploadStartPermissive(username, filename, token);
```

`UploadStartPermissive` would be a delegate returning a `Task`, and it would be supplied by the code implementing the library, the same way the logic for `EnqueueDownloadAction` works, basically.  @jackBonadies this should give you the level of control you describe without preventing other users from implementing other schemes.

I'm specifically looking at this because I'd like a way to set a number of slots per user class, and with just the one semaphore it wouldn't be possible.  In this case, `UploadStartPermissive()` would be implemented like...

```c#
permissive = (user, filename) => 
{
    if (<user is in group A>)
    {
        return GroupASemaphore.WaitAsync();
    }
    else if (<user is in group B>)
    {
        return GroupBSemaphore.WaitAsync();
    }

    // user isn't in a specific group, therefore no additional constraints other than global upload slot count
    return Task.CompletedTask;
}
```

I'm hesitant to double down on using semaphores to control queuing behavior, but this is based on how much memory a ton of enqueued uploads will take, so in the next few days I'll do some profiling to see if this is a legitimate fear.

Holding a read lock on files that are queued for uploading isn't the worst thing in the world, and losing the queue if the application restarts isn't either, since the official Soulseek do too.

**Comment by jpdillingham** [comment]
Another thing to note is that if it turns out starting uploads and using `await` to control traffic, it would probably be semantically correct to add an `EnqueueUploadAsync` that works the same way `EnqueueDownloadAsync` works.  In the case of uploads the code would need to return a `Task` after the state was changed to `Queued` and before any of the traffic control semaphores are awaited.

**Comment by jackBonadies** [comment]
Thanks for your work on this, I think the ability for the user to specify a delegate before the upload slot semaphore would be perfect for my use case (of FIFO or RR ordering).  A second delegate to call after upload (in the finally clause) in order to release the semaphore would also be useful.

I took a look at the basic queue #640, though I have not tried it out yet.  Its clean and definitely nice to have a default queuing mechanism.  Its another topic but I think a similar default implementation could be useful for the transfer governer for upload / download speed limits.  Reason being, that although one solution may not fit all (i.e. like you said speed limits for different groups, or global vs per-transfer limits), a default implementation would be useful for the common case.  And in my opinion it is a bit difficult for the end user to figure out the ideal governor to dictate transfer speed (i.e. do we add additional variables to keep track of the last uploaded amount, and the time per loop, to then do (last uploaded amount)/(time per loop + sleep time)=(desired speed limit) solving for sleep time?  or compare average speed of transfer to desired to increase / decrease wait time, which will be simpler, but maybe less exact and will take more loops for an appropriate wait time?).

**Comment by jpdillingham** [comment]
I've added three new delegates to `TransferOptions`:

```c#
Func<Transfer, CancellationToken, Task> acquireSlot
Action<Transfer> slotAcquired
Action<Transfer> slotReleased
```

I've adjusted the logic for uploads so that the following conditions must be met, in order:

* Per-user semaphore acquired
* Task obtained via `acquireSlot` invocation completed
* Global semaphore acquired

The practical effect of this is that `acquireSlot` is called the moment an upload is free to be started (barring a conflict with the global semaphore), so the order in which the `Task` returned by the delegate is completed will provide granular control over the order of uploads, while still enforcing the single upload per user constraint.

I implemented FIFO and Round Robin(-ish) queues in the example here: https://github.com/jpdillingham/Soulseek.NET/blob/start-permissive/examples/Web/api/Startup.cs#L489-L649

The code I ended up with is basically what you had originally described, so I think the API I'm exposing from the library is _probably_ sufficient.  I've created another issue #653 to expose Uploads and Downloads properties to give consumers a rough picture of what the queued transfers look like, which could help with decision making.

The only part of this that I don't really like is the following:

```c#
slotReleased: (tx) =>
{
    Console.WriteLine($"[UPLOAD SLOT RELEASED] [{username}/{filename}]");

    // return the semaphore to make way for the next upload
    ConcurrentUploadSemaphore.Release();

    // process the queue after a short delay, to give this user's next upload time to
    // make it into the waiting upload dictionary (if there is one)
    Task.Run(async () =>
    {
        await Task.Delay(100);
        _ = ProcessUploadQueue();
    });
}
```

Because transfers are isolated from one another and the traffic control logic relies on a semaphore, it is impossible to anticipate whether there's another upload waiting or what it would be within the body of `slotReleased`.  Without this delay the FIFO logic acts like round robin because the dictionary isn't aware of the next upload yet.  The only definitive solution I can come up with is to expose the internal per-user semaphore somehow, but that doesn't keep another thread from invoking `ProcessUploadQueue()` in the interim and having an upload sneak in.  I'll keep thinking about it but suggestions are welcome.

I think it probably would also be prudent to execute `ProcessUploadQueue` on a timer periodically, in case a transfer manages to catch the sync root at an odd time and gets stuck.

**Comment by jpdillingham** [comment]
Regarding the governance of speed.... I've honestly been putting it off because I'm not sure how I'll do it.  Calculating transfer speed is an inexact science, so governing it will be as well.  The governor function is called frequently inside of transfer loops which should give a lot of opportunity to average out errors.

I will probably end up adding a global speed governor within the library for both uploads and downloads as a start, since, like upload slots, it isn't really practical to offload that logic on every user of the library.  Along with this I'll have to implement a way to track global upload and download speeds and will make that available in the public API.

**Comment by jpdillingham** [comment]
@jackBonadies #649 has been merged into master.  I believe you're still using submodules, but if you've started pulling in the library from NuGet I can create a prerelease package so you can pull it in to test.

I did a bunch of testing around the timing of the release and next thread obtaining the per-user semaphore and in all cases the delay was less than 1ms.  I added a 10ms delay before calling `SlotReleased`, so if there is an upload waiting it should have invoked `AcquireSlot` prior to the `SlotReleased` from the previous upload being called.

Let me know what you think!

**Comment by jpdillingham** [comment]
I think there's enough functionality in this to release it for now.  I'll close this issue, but feel free to open another if you can think of any enhancements.
